src/megolm/inbound_group_session.rs

```diff
@@ -344,64 +329,36 @@ impl InboundGroupSession {
 
 #[derive(Serialize, Deserialize)]
 pub struct InboundGroupSessionPickle {
-    initial_ratchet: RatchetPickle,
-    latest_ratchet: RatchetPickle,
+    initial_ratchet: Ratchet,
     signing_key: Ed25519PublicKey,
     #[allow(dead_code)]
     signing_key_verified: bool,
 }
 
 impl InboundGroupSessionPickle {
-    pub fn unpickle(self) -> Result<InboundGroupSession, InboundGroupSessionUnpicklingError> {
-        self.try_into()
+    pub fn encrypt(self, pickle_key: &[u8; 32]) -> String {
+        pickle(&self, pickle_key)
     }
-}
 
-impl TryFrom<InboundGroupSessionPickle> for InboundGroupSession {
-    type Error = InboundGroupSessionUnpicklingError;
-
-    fn try_from(pickle: InboundGroupSessionPickle) -> Result<Self, Self::Error> {
-        Ok(Self {
-            initial_ratchet: pickle
-                .initial_ratchet
-                .try_into()
-                .map_err(InboundGroupSessionUnpicklingError::InvalidInitialRatchet)?,
-            latest_ratchet: pickle
-                .latest_ratchet
-                .try_into()
-                .map_err(InboundGroupSessionUnpicklingError::InvalidLatestRatchet)?,
-            signing_key: pickle.signing_key,
-            signing_key_verified: pickle.signing_key_verified,
-        })
+    pub fn from_encrypted(ciphertext: &str, pickle_key: &[u8; 32]) -> Result<Self, PickleError> {
+        unpickle(ciphertext, pickle_key)
     }
 }
 
-#[derive(Zeroize, Debug)]
-#[zeroize(drop)]
-pub struct InboundGroupSessionPickledJSON(String);
-
-impl InboundGroupSessionPickledJSON {
-    pub fn as_str(&self) -> &str {
-        &self.0
-    }
-
-    pub fn unpickle(self) -> Result<InboundGroupSession, InboundGroupSessionUnpicklingError> {
-        let pickle: InboundGroupSessionPickle = serde_json::from_str(&self.0)?;
-        pickle.unpickle()
-    }
-}
-
-impl AsRef<str> for InboundGroupSessionPickledJSON {
-    fn as_ref(&self) -> &str {
-        self.as_str()
+impl From<&InboundGroupSession> for InboundGroupSessionPickle {
+    fn from(session: &InboundGroupSession) -> Self {
+        session.pickle()
     }
 }
 
-impl Deref for InboundGroupSessionPickledJSON {
-    type Target = str;
-
-    fn deref(&self) -> &Self::Target {
-        self.as_ref()
+impl From<InboundGroupSessionPickle> for InboundGroupSession {
+    fn from(pickle: InboundGroupSessionPickle) -> Self {
+        Self {
+            initial_ratchet: pickle.initial_ratchet.clone(),
+            latest_ratchet: pickle.initial_ratchet,
```

Comment:
`latest_ratchet` gets reset to the `initial_ratchet` any time the `latest_ratchet` is advanced past the message index of the message that is being decrypted.

I don't think `latest_ratchet` plays part in the identity of the `InboundGroupSession`, we don't need to be byte for byte the same after a roundtrip.

Reply:
I'm aware of that, but we do have logic dependent on values of `initial_ratchet` and `latest_ratchet`. Introducing changes to `latest_ratchet` during a serialization cycle then complicates reasoning and introduces opportunities for unforeseen effects on that logic.

I guess I'm asking for a rationale on why we'd not want to save the state and remove this factor completely. Is it purely to make the pickle smaller or something else?

Reply:
> I'm aware of that, but we do have logic dependent on values of `initial_ratchet` and `latest_ratchet`. Introducing changes to `latest_ratchet` during a serialization cycle then complicates reasoning and introduces opportunities for unforeseen effects on that logic.

Could you clarify what you're talking about here? What kind of logic?

> I guess I'm asking for a rationale on why we'd not want to save the state and remove this factor completely. Is it purely to make the pickle smaller or something else?

Yes, we're wasting 128 bytes per inbound group session otherwise. I know that libolm pickles the memory as is and wastes a lot space that way. It's not something I'm keen on porting.

Reply:
> Could you clarify what you're talking about here? What kind of logic?

Sorry, I forgot to link. I was talking about e.g. `find_ratchet` (https://github.com/matrix-org/vodozemac/blob/bde1fbc96557d3b77ec1a6f92696468d155d2a4c/src/megolm/inbound_group_session.rs#L187) and `advance_to`.

> Yes, we're wasting 128 bytes per inbound group session otherwise. I know that libolm pickles the memory as is and wastes a lot space that way. It's not something I'm keen on porting.

I guess that's fair, though this way we need to recalculate things we already calculated instead, though that may be less of a problem. It's just something to be mindful of if we do decide to keep this; I don't feel too strongly either way.

Reply:
This is the code I was talking about as well, `latest_ratchet` will regularly be reset to `initial_ratchet`, so I don't think that there's anything to worry about. It is also set to `initial_ratchet` when we first create the the session.

In my view, `latest_ratchet` is only an optimization so that we don't need to advance from the start every time we decrypt a message, from the megolm spec:

>  They may also choose to cache calculated ratchet values,
but the decision of which ratchet states to cache is left to the application.